Propose clarification covering "persistence" across reboots and the term volatile

robshakir opened the pull request:
```
 * (M) docs/motivation.md
  - Clarify that dynamic configuration is not expected to persist
    across reloads.
  - Clarify that dynamic configuration is expected to persist across
    control plane switchovers.
```

robshakir committed:
Propose clarification covering "persistence" across reboots.

 * (M) docs/motivation.md
  - Clarify that dynamic configuration is not expected to persist
    across reloads.
  - Clarify that dynamic configuration is expected to persist across
    control plane switchovers.

dplore commented on docs/motivation.md:
```diff
@@ -277,6 +277,26 @@ assumes the existence of a coalescing queue on input. The client SHOULD NOT
 
+### Persistence of Configuration
+
+Like other control-plane protocols, the "dynamic" configuration of a device
```
The defniition of dynamic conflicts with dynamic in [bootz](https://github.com/openconfig/bootz/blob/a50e5d4599e34ef3fcfd546ad634a2f48ba54dc2/README.md?plain=1#L39) and [gnmi union_replace](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-union_replace.md#212-dynamic-configuration).  Could we call this ephemeral or volatile config?

robshakir replied:
I did suspect this would end up being overloaded.

Folks seemed to dislike "volatile" -- but it'd be my preference over "ephemeral". Ephemeral is defined, in rfc7921, thus:

>   Ephemeral state on a router is the state
>   that does not survive the reboot of a routing device or the reboot of
>   the software handling the I2RS software on a routing device. 

Now, if we say that this state should survive control-plane switchover events, then does this mean that it survives "the reboot of the software handling the i2rs software", and what does i2rs mean in this context.

So maybe I prefer volatile here, since it has less connotations in the industry. WDYT?

christopher-arista replied:
For what it's worth, internally we've been referring to this as "high-velocity input". As you note, "dynamic" is overloaded, and "config" has various other connotations for us (some baked-in assumptions about persistence, being part of "show running-config", etc).

Although admittedly I'm still somewhat partial to yolo-config :)

robshakir replied:
`yolo` is clearly the best choice, but maybe we go with `high-velocity` and just abbreviate this to `hv`? 

This would give us something like `/interfaces/interface/hv/oper-status` for example.

robshakir replied:
On second thoughts, this is a bit cryptic -- so maybe just `volatile`.

We should put together an updated version of `draft-openconfig-netmod-opstate` that has a clear definition of the types of state that exist within OpenConfig at this point in time -- including those that are defined in the `bootz` document.

dplore replied:
+1 to `volatile`.  It doesn't collide with other commonly used terms and does imply semantics of 'temporary and changes often, quickly, etc.'.  :)

dplore commented on docs/motivation.md:
```diff
@@ -277,6 +277,26 @@ assumes the existence of a coalescing queue on input. The client SHOULD NOT
+
+Like other control-plane protocols, the "dynamic" configuration of a device
+is not expected to be persisted across device reboots. Data that is written
+via gNMI to a `config` path is expected to be persisted after each transaction.
```
and across reloads

robshakir replied:
Persisted after each transaction means that it's persisted across reloads. Not sure this clarification is needed. I've made a note that this storage it is persisted to is non-volatile which seems like the only clarification needed.

dplore commented on docs/motivation.md:
```diff
@@ -277,6 +277,26 @@ assumes the existence of a coalescing queue on input. The client SHOULD NOT
+to the `dynamic` path according to some runtime condition of the wider network
+system.
+
+Note that `persistence` within the service definition refers to persistence
+across RPCs (i.e., clients connecting or disconnecting) not across device
+reboots.
```
```suggestion
Note that `persistence` within the gsii service definition refers to persistence
across RPCs (i.e., clients connecting or disconnecting) not across device
reboots.
```

robshakir committed:
Address review comments.

 * (M) docs/motivation.md
   - Rename `dynamic` to `volatile` to ensure that do not overload
     terms defined in the bootz specification.
   - Minor clarifications.

robshakir committed:
Merge branch 'main' into clarifications

dplore commented on docs/motivation.md:
```diff
@@ -130,7 +130,7 @@ can be triggered by an NOS-external application.
 ## Requirements
 
-* **[R1]** The elements that are provided as a dynamic parameter should be _in
+* **[R1]** The elements that are able to be changed more rapidly should be _in
```
here and throughout, can we standardize on terms for the configuration stores:
volatile (as per gsii definition)
dynamic (as per bootz definition)
applied (as per the draft-openconfig-netmod-opstate-01)

So for example:

```suggestion
* **[R1]** The `applied` set of configuration items is the union of `volatile` configuration items and  `dynamic` configuration items on the device.   Removing the volatile set does not affect the dynamic set.

* **[R2]** volatile configuration items override dynamic configuration items.
```

robshakir replied:
Given that these requirements are described independently of the solution -- I don't think this makes sense to do. This document is meant to sketch out the initial solution and similarly to gRIBI, I propose to write a specification *after* iterating on the design.

This requirement as you have suggested is very specific to an individual solution.

dplore commented on docs/motivation.md:
```diff
@@ -235,44 +234,56 @@ via this new API – since clearly there are some cases where strong validation
+described in this document). Therefore, we propose that where payloads are
+generated for the `Modify` RPC, we give these a new place within the OpenConfig
+schema, particularly, defining a new
+`volatile`/`ephemeral` subtree to correspond with the `config` and
```
Maybe the design decisions for modeling should be outside the scope of the motivation doc?

That said, would an annotation be more appropriate?

robshakir replied:
They can be moved, yes.

I'm not sure how an annotation would work here -- what are you proposing? There are other PRs that cover how the content of these RPCs are proposed to be generated. PTAL at #6.

dplore replied:
The volatile container could have an annotation so there is a property in yang identifying the node as volatile. 
It might be useful so yang aware code can identify volatile leafs programmatically versus relying on a "magic" container name.   (Although I think using "volatile" as a container name also makes sense from a style perspective.  I would do both the annotation an the container name)  

An analogous entity is using the [yang config statement](https://datatracker.ietf.org/doc/html/rfc6020#section-7.19.1), but also as an OC style guide, we use a container named 'config'.

robshakir replied:
Ah, I see -- yes, annotating as well as a container name makes sense. I think we must have a container name for consistency. The alternate suggestion of having an annotation on an existing `config` leaf would be possible but add more complexity of magic in the tooling to know that such an annotation means that there is another path for the leaf.

dplore replied:
Ok, agreed that we should have both:  a container and an annotation such as `config-volatile`, as this is the most explicit.

brianneville commented:
can this commit message/title be updated to reflect that this commit also clarifies the term`volatile` in reference to this rapidly-changing config?

dplore commented:
> can this commit message/title be updated to reflect that this commit also clarifies the term`volatile` in reference to this rapidly-changing config?

I updated the title
